### PR TITLE
Add site admin banner warning when exceeding Cody limits

### PR DESCRIPTION
**Pull request by eseliger**
This PR adds a mechanism that syncs the rate limits in the background, stores the usage in redis every 30 minutes, and displays a banner warning when the rate limits are being exceeded:
- Info at 75%+
- Warning at 90%+
- Error at 100%

This PR also moves the code for talking to Cody Gateway to get usage data into the common codebase out of app, so that we get all the data exposed through graphql for every instance. This may be useful for debugging, or implementing additional APIs to visualize limits.

Screenshot 2023-08-01 at 00 19 49@2x

## Test plan

Verified banners show correctly.

### CODE REVIEW

File: cmd/worker/internal/codygateway/usageworker.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+						ttl = timeToReset
+					}
+				}
+				if err := redispool.Store.SetEx(fmt.Sprintf("%s:%s", codygateway.CodyGatewayUsageRedisKeyPrefix, string(l.Feature)), ttl, l.PercentUsed()); err != nil {
+					j.logger.Error("failed to store rate limit usage for cody gateway", log.Error(err))
+				}
+			}
+		}
+
+		// Run once on init:
```
Comment (unknwon) [inline cmd/worker/internal/codygateway/usageworker.go:85]:
```suggestion
		// Run once on init
```

fat finger?

File: cmd/worker/internal/codygateway/usageworker.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+			limits, err := cgc.GetLimits(j.ctx)
+			if err != nil {
+				j.logger.Error("failed to get cody gateway limits", log.Error(err))
+				return
+			}
+
+			for _, l := range limits {
+				ttl := redisTTLMinutes * 60
+				// Make sure the expiry will happen at least every redisTTLMinutes,
+				// but when the limit actually expires otherwise.
```
Comment (unknwon) [inline cmd/worker/internal/codygateway/usageworker.go:69]: 🤔 not sure I understanding the relation between the "but ..." part and the first half of the sentence.

File: cmd/worker/internal/codygateway/usageworker.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+				if l.Expiry != nil {
+					timeToReset := int(time.Until(*l.Expiry).Seconds())
+					if timeToReset == 0 {
```
Comment (unknwon) [inline cmd/worker/internal/codygateway/usageworker.go:72]: nit
```suggestion
					if timeToReset <= 0 {
```

File: internal/codygateway/client.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	Feature Feature
+
+	IntervalLimit int64      `json:"limit"`
+	IntervalUsage int64      `json:"usage"`
+	TimeInterval  string     `json:"interval"`
+	Expiry        *time.Time `json:"expiry"`
+}
+
+func (rl LimitStatus) PercentUsed() int {
+	if rl.IntervalLimit == 0 {
```
Comment (unknwon) [inline internal/codygateway/client.go:30]: nit
```suggestion
	if rl.IntervalLimit <= 0 {
```

File: internal/codygateway/client.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+		return 100
+	}
+	return int(math.Ceil(float64(rl.IntervalUsage) / float64(rl.IntervalLimit) * 100))
+}
+
+type Client interface {
+	GetLimits(ctx context.Context) ([]LimitStatus, error)
+}
+
+func NewClientFromSiteConfig(cli httpcli.Doer) Client {
```
Comment (unknwon) [inline internal/codygateway/client.go:40]: I think we should either have docstring saying returning `nil` meaning no client configured or use `(Client, bool)` as return values, for a more idiomatic Go pattern.

**Review by unknwon — APPROVED** [review]
👍

**Review by bobheadxi — APPROVED** [review]
Approve to unblock and overall LGTM, but this exposes a lot more about Cody Gateway within Sourcegraph itself than I might have anticipated we would, particularly in the GraphQL layer

I might have introduced something such as:

```
site {
   codyLLMConfiguration {
      status {
         summary
      }
   }
}
```

and left each provider type to implement a way of summarizing its rate limits, e.g. if one day we expand BYOK(ey)/BYOLLM we can have ways of summarizing each type of LLM provider. Not a huge deal, just my 2c

**Comment by eseliger** [comment]
> Approve to unblock and overall LGTM, but this exposes a lot more about Cody Gateway within Sourcegraph itself than I might have anticipated we would, particularly in the GraphQL layer

That's an interesting take, I actually basically just took what App did and exposed it for non-app instances as well, no much further thinking here. I see where you're going with that, and it probably makes sense to have this be more generic per provider. Something like:

```go
type Provider interface { Name() string }
type CompletionsProvider interface { Complete(...) ... Stream(...) ... }
type EmbeddingsProvider interface { Embed(...) ... }
type RateLimitedProvider interface { Status(feature) FormatAlert(usage) string }
```

might be interesting to look at 🤔 Then, the worker can create a provider for each of the supported features, and make sure that we get the limits for _each_ of the services, potentially also providing rendering for the message or such. 

That does feel pretty compelling to me actually, though I wonder how far we should go initially with this given we'll backport this and it'll turn the code more upside down than might be easy to port. I think we can use the code in here to later build out this generic feature set pretty easily - how do you feel about that?

**Comment by bobheadxi** [comment]
> That's an interesting take, I actually basically just took what App did and exposed it for non-app instances as well

Yeah, it does look like some of the GraphQL stuff got added earlier so extending it kind of makes sense for now if we want to add this particular feature 

> That does feel pretty compelling to me actually, though I wonder how far we should go initially with this given we'll backport this and it'll turn the code more upside down than might be easy to port. I think we can use the code in here to later build out this generic feature set pretty easily - how do you feel about that?

Is there an urgent ask for this feature? In #cody-usage-notifications so far there does not seem to have been many instances of actual Sourcegraph instances hitting Cody Gateway rate limits. No problem if we have a requirement to meet here, but IMO if there's no rush then maybe we can use this opportunity to get this right first time around?

In other words - App users encounter limits much more frequently so I can understand the urgency there, but this seems less urgent from what I've noticed, and with the difficulty of backports it seems beneficial to avoid landing too much stuff that will get changed soon-ish.

Either way, I defer to your judgement here 😁

**Comment by eseliger** [comment]
Talked through this on slack, as a temporary solution, we're going with this.
